<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the `r Biocpkg("iSEE")` interface 
 • iSEE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the `r Biocpkg(" isee interface>
<meta property="og:description" content="iSEE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iSEE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic.html">An introduction to the `r Biocpkg("iSEE")` interface 
</a>
    </li>
    <li>
      <a href="../articles/bigdata.html">How to use `r Biocpkg("iSEE")` with big data
</a>
    </li>
    <li>
      <a href="../articles/configure.html">Configuring `r Biocpkg("iSEE")` apps
</a>
    </li>
    <li>
      <a href="../articles/custom.html">Deploying custom panels in the `iSEE` interface</a>
    </li>
    <li>
      <a href="../articles/ecm.html">Describing the `ExperimentColorMap` class</a>
    </li>
    <li>
      <a href="../articles/links.html">Sharing information across `r Biocpkg("iSEE")` panels
</a>
    </li>
    <li>
      <a href="../articles/voice.html">Controlling the `iSEE` interface using speech recognition</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iSEE/iSEE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="basic_files/font-awesome-5.1.0/css/all.css" rel="stylesheet">
<link href="basic_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
<p>An introduction to the <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
interface</p>
                        <h4 data-toc-skip class="author">Kevin
Rue-Albrecht</h4>
            <address class="author_afil">
      MRC WIMM Centre for Computational Biology, University of Oxford,
Oxford, OX3 9DS,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Federico
Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), MainzCenter for Thrombosis and Hemostasis (CTH),
Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                              <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
            <address class="author_afil">
      Friedrich Miescher Institute for Biomedical Research, Basel,
SwitzerlandSIB Swiss Institute of
Bioinformatics<br><a class="author_email" href="mailto:#"></a><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Aaron
Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">1 June 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/iSEE/iSEE/blob/HEAD/vignettes/basic.Rmd" class="external-link"><code>vignettes/basic.Rmd</code></a></small>
      <div class="hidden name"><code>basic.Rmd</code></div>

    </div>

    
    
<p><strong>Compiled date</strong>: 2022-06-01</p>
<p><strong>Last edited</strong>: 2020-04-20</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<p><img src="../../../_temp/Library/iSEE/www/iSEE.png" width="50%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> is a <a href="http://bioconductor.org" class="external-link">Bioconductor</a> package that provides an
interactive Shiny-based graphical user interface for exploring data
stored in <code>SummarizedExperiment</code> objects <span class="citation">(Rue-Albrecht et al. 2018)</span>. Instructions to
install the package are available <a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">here</a>. Once
installed, the package can be loaded and attached to your current
workspace as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></span><span class="op">)</span></code></pre></div>
<p>If you have a <code>SummarizedExperiment</code> object<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> named <code>se</code>,
you can launch an <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> app by
running:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/iSEE.html">iSEE</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></code></pre></div>
<p>In this vignette, we demonstrate this process using the
<code>allen</code> single-cell RNA-seq data set from the <em><a href="https://bioconductor.org/packages/3.15/scRNAseq" class="external-link">scRNAseq</a></em>
package. However, if you want to start playing with the app immediately,
you can simply run:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">iSEE</span>, ask<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="setting-up-the-data">Setting up the data<a class="anchor" aria-label="anchor" href="#setting-up-the-data"></a>
</h2>
<p>The <code>allen</code> data set contains expression values for 379
cells from the mouse visual cortex <span class="citation">(Tasic et al.
2016)</span>, and can be loaded directly by calling
<code><a href="https://rdrr.io/pkg/scRNAseq/man/ReprocessedData.html" class="external-link">ReprocessedAllenData()</a></code> and specifying the value for the
<code>assays</code> parameter. To begin with, we assign the output of
this call to an <code>sce</code> object and inspect it.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ReprocessedData.html" class="external-link">ReprocessedAllenData</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"tophat_counts"</span><span class="op">)</span>   <span class="co"># specifying the assays to speed up the example</span>
<span class="va">sce</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 20816 379 </span>
<span class="co">#&gt; metadata(2): SuppInfo which_qc</span>
<span class="co">#&gt; assays(1): tophat_counts</span>
<span class="co">#&gt; rownames(20816): 0610007P14Rik 0610009B22Rik ... Zzef1 Zzz3</span>
<span class="co">#&gt; rowData names(0):</span>
<span class="co">#&gt; colnames(379): SRR2140028 SRR2140022 ... SRR2139341 SRR2139336</span>
<span class="co">#&gt; colData names(22): NREADS NALIGNED ... Animal.ID passes_qc_checks_s</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; mainExpName: endogenous</span>
<span class="co">#&gt; altExpNames(1): ERCC</span></code></pre></div>
<p>As provided, the <code>sce</code> object contains raw data and a
number of quality control and experimental cell annotations, all
available in <code>colData(sce)</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;  [1] "NREADS"                       "NALIGNED"                    </span>
<span class="co">#&gt;  [3] "RALIGN"                       "TOTAL_DUP"                   </span>
<span class="co">#&gt;  [5] "PRIMER"                       "PCT_RIBOSOMAL_BASES"         </span>
<span class="co">#&gt;  [7] "PCT_CODING_BASES"             "PCT_UTR_BASES"               </span>
<span class="co">#&gt;  [9] "PCT_INTRONIC_BASES"           "PCT_INTERGENIC_BASES"        </span>
<span class="co">#&gt; [11] "PCT_MRNA_BASES"               "MEDIAN_CV_COVERAGE"          </span>
<span class="co">#&gt; [13] "MEDIAN_5PRIME_BIAS"           "MEDIAN_3PRIME_BIAS"          </span>
<span class="co">#&gt; [15] "MEDIAN_5PRIME_TO_3PRIME_BIAS" "driver_1_s"                  </span>
<span class="co">#&gt; [17] "dissection_s"                 "Core.Type"                   </span>
<span class="co">#&gt; [19] "Primary.Type"                 "Secondary.Type"              </span>
<span class="co">#&gt; [21] "Animal.ID"                    "passes_qc_checks_s"</span></code></pre></div>
<p>Then, we normalize the expression values with <em><a href="https://bioconductor.org/packages/3.15/scater" class="external-link">scater</a></em>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, exprs_values<span class="op">=</span><span class="st">"tophat_counts"</span><span class="op">)</span></code></pre></div>
<p>Next, we apply PCA and <em>t</em>-SNE to generate two low-dimensional
representations of the cells. The dimensionality reduction results are
stored in <code>reducedDim(sce)</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>
<span class="co">#&gt; [1] "PCA"  "TSNE"</span></code></pre></div>
<p>At this point, the <code>sce</code> object does not contain any
annotations for the rows (i.e., features) in the data set. Thus, to
prepare a fully-featured example application, we also add some gene
metadata to the <code>rowData</code> related to the mean-variance
relationship in the data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">mean_log</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">var_log</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span></code></pre></div>
<p>It is important to note that <code>iSEE</code> relies primarily on
precomputed values stored in the various slots of objects derived from
the <code>SummarizedExperiment</code> class<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. This allows users to
visualize any metrics of interest, but also requires these to be
calculated and added to the object before the initialization of the app.
That said, it is straightforward to iteratively explore a precomputed
object, take notes of new metrics to compute, close the app, store new
results in the <code>SummarizedExperiment</code> object, and launch a
new app using the updated object.</p>
</div>
<div class="section level2">
<h2 id="launching-the-interface">Launching the interface<a class="anchor" aria-label="anchor" href="#launching-the-interface"></a>
</h2>
<p>To begin the exploration, we create an <code>iSEE</code> app with the
<code>SingleCellExperiment</code> object generated above. In its
simplest form, the <code>iSEE</code> function only requires the input
object. However, <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
applications can be extensively <a href="https://bioconductor.org/packages/3.15/iSEE/vignettes/configure.html" class="external-link">reconfigured</a>
using a number of optional arguments to the <code>iSEE</code>
function.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iSEE.html">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></code></pre></div>
<p>The <code>runApp</code> function launches the app in our browser.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">app</span><span class="op">)</span></code></pre></div>
<p><img src="screenshots/basic-demo.png" width="488"></p>
<p>By default, the app starts with a dashboard that contains one panel
or table of each type. By opening the collapsible panels named “Data
parameters”, “Visual parameters”, and “Selection parameters” under each
plot, we can control the content and appearance of each panel.</p>
<p>Now, look in the upper right corner for a question mark icon
(<i class="fa fa-question-circle"></i>), and click on the hand button
(<i class="fa fa-hand-o-right"></i>) for an introductory tour. This will
perform an interactive tour of the app, based on the <em><a href="https://CRAN.R-project.org/package=rintrojs" class="external-link">rintrojs</a></em>
package <span class="citation">(Ganz 2016)</span>. During this tour, you
will be taken through the different components of the <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> user
interface and learn the basic usage mechanisms by doing small actions
guided by the tutorial: the highlighted elements will be responding to
your actions, while the rest of the UI will be shaded. You can move
forward and backward along the tour by clicking on the “Next”/“Back”
buttons, or also using the arrow keys. You can even jump to a particular
step by clicking on its circle. To exit the tour, either click on
“Skip”, or simply click outside of the highlighted UI element.</p>
<p>Once you are done generating plots, click on the export icon
(<i class="fa fa-download fa-1g"></i>) in the upper right corner, and
click on the magic wand button (<i class="fa fa-magic"></i>) to display
R code that you can export and directly re-use in your R session. This
will open a modal popup where the R code used to generate the plots is
displayed in a <em><a href="https://CRAN.R-project.org/package=shinyAce" class="external-link">shinyAce</a></em>-based
text editor. Select parts or all of it to copy-and-paste it into your
analysis script/Rmarkdown file. However, note that the order in which
the code blocks are reported is important if you have linked panels to
one another, as the panels sending point selections must be executed
before those that receive the corresponding selection.</p>
</div>
<div class="section level2">
<h2 id="description-of-the-user-interface">Description of the user interface<a class="anchor" aria-label="anchor" href="#description-of-the-user-interface"></a>
</h2>
<div class="section level3">
<h3 id="header">Header<a class="anchor" aria-label="anchor" href="#header"></a>
</h3>
<p>The layout of the <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> user
interface uses the <em><a href="https://CRAN.R-project.org/package=shinydashboard" class="external-link">shinydashboard</a></em>
package. The dashboard header contains four dropdown menus.</p>
<p>The first icon in the dashboard header is the “Organization” menu,
which is identified by an icon displaying multiple windows
(<i class="fa fa-object-ungroup"></i>). This menu contains two
items:</p>
<ul>
<li>The “Organize panels” button. Click on this button to open a modal
window that contains:
<ul>
<li>A selectize input to add, remove, and reorder panels in the main
interface.</li>
<li>Two inputs to control the width and height, respectively, of each
panel selected above. Note that panel width is defined in integer grid
units between 1 and 12 (see <code><a href="https://rdrr.io/pkg/shiny/man/column.html" class="external-link">?shiny::column</a></code>). In contrast,
panel height is defined in pixels (see
<code><a href="https://rdrr.io/pkg/shiny/man/plotOutput.html" class="external-link">?shiny::plotOutput</a></code>).</li>
</ul>
</li>
<li>The “Examine panel chart” functionality, identified by a chain icon
(<i class="fa fa-chain"></i>). Click on this button to obtain a graph
representation of the existing links and point selections among your
visible plot and table panels. Every panel is represented by a node
coded with the same color as in the app. This can be very useful in
sessions that include a large number of panels, to visualize the
relationship structure between the various panels that send and receive
selections of data points.</li>
</ul>
<p>The second icon in the dashboard header is the “Export” dropdown
menu, which is identified by a download icon
(<i class="fa fa-download fa-1g"></i>) and contains:</p>
<ul>
<li>The “Download panel output” functionality, identified by the
download icon (<i class="fa fa-download fa-1g"></i>). Here, you can
download a zip folder with the currently displayed panel content
(high-resolution figures and table contents as csv files).</li>
<li>The “Extract the R code”, functionality
(<i class="fa fa-magic"></i>). At any point during your live session,
you might want to record the code that reproduces exactly the state of
each plot. Clicking on this button opens a modal popup window, with a
<em><a href="https://CRAN.R-project.org/package=shinyAce" class="external-link">shinyAce</a></em>-based
text editor, where the code is formatted and displayed with syntax
highlighting. You can copy the code to the clipboard by selecting the
text (please do include the initial lines and the
<code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> commands for best tracking of your
environment), and store it in your analysis report/script. This code can
then be further edited to finalize the plots (e.g., for
publication).</li>
<li>The “Display panel settings” (<i class="fa fa-clipboard"></i>)
button, which lets you export the code defining the state of the current
panels in the interface. This is useful if you want to preconfigure an
<em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
instance to start in the current state, rather than with the default set
of panels.</li>
</ul>
<p>The “Documentation” dropdown menu is accessible through the question
mark icon (<i class="fa fa-question-circle"></i>), which contains:</p>
<ul>
<li>The button to start an interactive tour
(<i class="fa fa-hand-o-right"></i>) of <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>, which
allows users to learn the basic usage mechanisms by doing. During a
tour, the highlighted elements respond to the user’s actions, while the
rest of the UI is shaded.</li>
<li>The button to “Open the vignette” (<i class="fa fa-book"></i>),
which displays the <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
vignette, either available on your system or accessed at the webpage of
the package on the Bioconductor project site. In the latter case, the
vignette will refer to the current release or development version,
according to the version of the package installed on your system).</li>
</ul>
<p>The “Additional Information” dropdown menu is accessible through the
information icon (<i class="fa fa-info"></i>), and contains:</p>
<ul>
<li>The “About this session” button
(<i class="fa fa-window-maximize"></i>), which reports the output of the
<code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> function in a modal popup window. This is
particularly useful for reproducing or reporting the environment,
especially when reporting errors or unexpected behaviors.</li>
<li>The “About iSEE” button (<i class="fa fa-heart"></i>) shows the
information on the development team, licensing and citation information
for the <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
package. You can follow the development of the package by checking the
GitHub repository (<a href="https://github.com/iSEE/iSEE" class="external-link uri">https://github.com/iSEE/iSEE</a>), where new functionality
will be added. Well-considered suggestions in the form of issues or pull
requests are welcome.</li>
</ul>
</div>
<div class="section level3">
<h3 id="body">Body<a class="anchor" aria-label="anchor" href="#body"></a>
</h3>
<div class="section level4">
<h4 id="overview-of-panel-types">Overview of panel types<a class="anchor" aria-label="anchor" href="#overview-of-panel-types"></a>
</h4>
<p>The main element in the body of <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> is the
combination of panels, generated (and optionally linked to one another)
according to your actions. There are currently eight standard panel
types that can be generated with <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>:</p>
<ul>
<li>Reduced dimension plot</li>
<li>Column data table</li>
<li>Column data plot</li>
<li>Feature assay plot</li>
<li>Row data table</li>
<li>Row data plot</li>
<li>Sample assay plot</li>
<li>Complex heatmap</li>
</ul>
<p>In addition, custom panel types can be defined as described in a <a href="https://bioconductor.org/packages/3.15/iSEE/vignettes/custom.html" class="external-link">separate
dedicated vignette</a>. The panels and models in the <em><a href="https://bioconductor.org/packages/3.15/iSEEu" class="external-link">iSEEu</a></em>
package provide additional flexibility.</p>
<p>For each standard plot panel, three different sets of parameters will
be available in collapsible boxes:</p>
<ul>
<li>“Data parameters”, to control parameters specific to each type of
plot.</li>
<li>“Visual parameters”, to specify parameters that will determine the
aspect of the plot, in terms of coloring, point features, and more
(e.g., legend placement, font size).</li>
<li>“Selection parameters” to control the incoming point selection and
link relationships to other plots.</li>
</ul>
</div>
<div class="section level4">
<h4 id="reduced-dimension-plots">Reduced dimension plots<a class="anchor" aria-label="anchor" href="#reduced-dimension-plots"></a>
</h4>
<p>If a <code>SingleCellExperiment</code> object is supplied to the
<code>iSEE</code> function, reduced dimension results are extracted from
the <code>reducedDim</code> slot. Examples include low-dimensional
embeddings from principal components analysis (PCA) or
<em>t</em>-distributed stochastic neighbour embedding (<em>t</em>-SNE)
<span class="citation">(Van der Maaten and Hinton 2008)</span>. These
results are used to construct a two-dimensional <em>Reduced dimension
plot</em> where each point is a sample, to facilitate efficient
exploration of high-dimensional datasets. The “Data parameters” control
the <code>reducedDim</code> slot to be displayed, as well as the two
dimensions to plot against each other.</p>
<p>Note that this builtin panel does not compute reduced dimension
embeddings; they must be precomputed and available in the object
provided to the <code>iSEE</code> function. Nevertheless, custom panels
- such as the <em><a href="https://bioconductor.org/packages/3.15/iSEEu" class="external-link">iSEEu</a></em>
<code>DynamicReducedDimensionPlot</code> can be developed and used to
enable such features.</p>
</div>
<div class="section level4">
<h4 id="coldataplot">Column data plots<a class="anchor" aria-label="anchor" href="#coldataplot"></a>
</h4>
<p>A <em>Column data plot</em> visualizes sample metadata stored in the
<code>SummarizedExperiment</code> column metadata. Different fields can
be used for the x- and y-axes by selecting appropriate values in the
“Data parameters” box. This plot can assume various forms, depending on
the nature of the data on the x- and y-axes:</p>
<ul>
<li>If the y-axis is continuous and the x-axis is categorical, violin
plots are generated (grouped by the x-axis factor).</li>
<li>If the y-axis is categorical and the x-axis is continuous,
horizontal violin plots are generated (grouped by the y-axis
factor).</li>
<li>If both axes are continuous, a scatter plot is generated. This
enables the use of contours that are overlaid on top of the plot, check
the <code>"Other"</code> box to see the available options.</li>
<li>If both axes are categorical, a plot of squares (Hinton plot) is
generated where the area of each square is proportional to the number of
samples within each combination of factor levels.</li>
</ul>
<p>Note that an x-axis setting of “None” is considered to be categorical
with a single level.</p>
</div>
<div class="section level4">
<h4 id="feature-assay-plots">Feature assay plots<a class="anchor" aria-label="anchor" href="#feature-assay-plots"></a>
</h4>
<p>A <em>Feature assay plot</em> visualizes the assayed values (e.g.,
gene expression) for a particular feature (e.g., gene) across the
samples on the y-axis. This usually results in a (grouped) violin plot,
if the x-axis is set to <code>"None"</code> or a categorical variable;
or a scatter plot, if the x-axis is another continuous variable<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>.</p>
<p>Gene selection for the y-axis can be achieved by using a <em>linked
row data table</em> in another panel. Clicking on a row in the table
automatically changes the assayed values plotted on the y-axis.
Alternatively, the row name can be directly entered as text that
corresponds to an entry of <code>rownames(se)</code><a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>.</p>
<p>The x-axis covariate can also be selected from the plotting
parameters. This can be <code>"None"</code>, sample metadata, or the
assayed values of another feature (also identified using a linked table
or via text). The measurement units are selected as one of the
<code>assays(se)</code>, which is applied to both the X and Y axes.</p>
<p>Obviously, any other assayed value for any feature can be visualized
in this manner, not limited to the expression of genes. The only
requirement for this type of panel is that the observations can be
stored as a matrix in the <code>SummarizedExperiment</code> object.</p>
</div>
<div class="section level4">
<h4 id="row-data-plots">Row data plots<a class="anchor" aria-label="anchor" href="#row-data-plots"></a>
</h4>
<p>A <em>Row data plot</em> allows the visualization of information
stored in the <code>rowData</code> slot of a
<code>SummarizedExperiment</code> object. Its behavior mirrors the
implementation for the <em>Column data plot</em>, and correspondingly
this plot can assume various forms depending on whether the data are
categorical or continuous.</p>
</div>
<div class="section level4">
<h4 id="sample-assay-plots">Sample assay plots<a class="anchor" aria-label="anchor" href="#sample-assay-plots"></a>
</h4>
<p>A <em>Sample assay plot</em> visualizes the assayed values (e.g.,
gene expression) for a particular sample (e.g., cell) across the
features on the y-axis.</p>
<p>This usually results in a (grouped) violin plot, if the x-axis is set
to <code>"None"</code> or a categorical variable (e.g., gene biotype);
or a scatter plot, if the x-axis is another continuous variable.</p>
<p>Notably, the x-axis covariate can also be set to:</p>
<ul>
<li>A discrete row data covariates (e.g., gene biotype), to stratify the
distribution of assayed values</li>
<li>A continuous row data covariate (e.g., count of cells expressing
each gene)</li>
<li>Another sample, to visualize and compare the assayed values in any
two samples.</li>
</ul>
</div>
<div class="section level4">
<h4 id="row-data-tables">Row data tables<a class="anchor" aria-label="anchor" href="#row-data-tables"></a>
</h4>
<p>A <em>Row data table</em> contains the values of the
<code>rowData</code> slot for the
<code>SingleCellExperiment</code>/<code>SummarizedExperiment</code>
object. If none are available, a column named <code>Present</code> is
added and set to <code>TRUE</code> for all features, to avoid issues
with <code><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">DT::datatable</a></code> and an empty <code>DataFrame</code>.
Typically, these tables are used to link to other plots to determine the
features to use for plotting or coloring. However, they can also be used
to retrieve gene-specific annotation on the fly by specifying the
<code>annotFun</code> parameter, e.g. using the
<code>annotateEntrez</code> or <code>annotateEnsembl</code> functions,
provided in <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>.
Alternatively, users can create a customized annotation function; for
more details on this, please consult the manual pages
<code>?annotateEntrez</code> and <code>?annotateEnsembl</code>.</p>
</div>
<div class="section level4">
<h4 id="column-data-tables">Column data tables<a class="anchor" aria-label="anchor" href="#column-data-tables"></a>
</h4>
<p>A <em>Column data table</em> contains the values of the
<code>colData</code> slot for the
<code>SingleCellExperiment</code>/<code>SummarizedExperiment</code>
object. Its behavior mirrors the implementation for the <em>Row data
table</em>. Correspondingly, if none are available, a column named
<code>Present</code> is added and set to <code>TRUE</code> for all
samples, to avoid issues with <code><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">DT::datatable</a></code> and an empty
<code>DataFrame</code>. Typically, these tables are used to link to
other plots to determine the samples to use for plotting or
coloring.</p>
</div>
<div class="section level4">
<h4 id="heat-maps">Heat maps<a class="anchor" aria-label="anchor" href="#heat-maps"></a>
</h4>
<p><em>Heat map</em> panels provide a compact overview of the data for
multiple features in the form of color-coded matrices. These correspond
to the <code>assays</code> stored in the
<code>SummarizedExperiment</code> object, where features (e.g., genes)
are the rows and samples are the columns.</p>
<p>User can select features (rows) to display from the selectize widget
(which supports autocompletion), or also via other panels, like row data
plots or row data tables. In addition, users can rapidly import custom
lists of feature names using a modal popup that provides an <a href="https://ace.c9.io/" class="external-link">Ace editor</a> where they can directly type of
paste feature names, and a file upload button that accepts text files
containing one feature name per line. Users should remember to click the
“Apply” button before closing the modal, to update the heat map with the
new list of features.</p>
<p>The “Suggest feature order” button clusters the rows, and also
rearranges the elements in the selectize according to the clustering. It
is also possible to choose which assay type is displayed
(<code>"logcounts"</code> being the default choice, if available).
Samples in the heat map can also be annotated, simply by selecting
relevant column metadata. A zooming functionality is also available,
restricted to the y-axis (i.e., allowing closer inspection on the
individual features included).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="description-of-isee-functionality">Description of iSEE functionality<a class="anchor" aria-label="anchor" href="#description-of-isee-functionality"></a>
</h2>
<div class="section level3">
<h3 id="coloring-plots-by-sample-attributes">Coloring plots by sample attributes<a class="anchor" aria-label="anchor" href="#coloring-plots-by-sample-attributes"></a>
</h3>
<p>Column-based plots are the reduced dimension, feature assay and
column data plots, where each data point represents a sample. Here, data
points can be colored in different ways:</p>
<ul>
<li>The default is no color scheme (<code>"None"</code> in the radio
button). This results in data points of a constant user-specified
color.</li>
<li>Any column of <code>colData(se)</code> can be used. The plot
automatically adjusts the scale to use based on whether the chosen
column is continuous or categorical.</li>
<li>The assay values of a particular feature in each sample can be used.
The feature can be chosen either via a linked row table or selectize
input (as described for the <em>Feature assay plot</em> panel). Users
can also specify the <code>assays</code> from which values are
extracted.</li>
<li>The identity of a particular sample can be used, which will be
highlighted on the plot in a user-specified color. The sample can be
chosen either via a linked column table or via a selectize input.</li>
</ul>
<p>For row-based plots (i.e., the sample assay and row data plots), each
data point represents a feature. Like the column-based plots, data
points can be colored by:</p>
<ul>
<li>
<code>"None"</code>, yielding data points of fixed color.</li>
<li>Any column of <code>rowData(se)</code>.</li>
<li>The identity of a particular <em>feature</em>, which is highlighted
in the user-specified color.</li>
<li>Assay values for a particular <em>sample</em>.</li>
</ul>
<p>Fine control of the color maps is possible through the
<code>ExperimentColorMap</code> class, see <a href="https://bioconductor.org/packages/3.15/iSEE/vignettes/ecm.html" class="external-link">this
vignette</a> for more details.</p>
</div>
<div class="section level3">
<h3 id="controlling-point-aesthetics">Controlling point aesthetics<a class="anchor" aria-label="anchor" href="#controlling-point-aesthetics"></a>
</h3>
<p>Data points can be set to different shapes according to categorical
factors in <code>colData(se)</code> (for column-based plots) or
<code>rowData(se)</code> (for row-based plots). This is achieved by
checking the <code>"Shape"</code> box to reveal the shape-setting
options. The size and opacity of the data points can be modified via the
options available by checking the <code>"Point"</code> box. This may be
useful for aesthetically pleasing visualizations when the number of
points is very large or small.</p>
</div>
<div class="section level3">
<h3 id="faceting">Faceting<a class="anchor" aria-label="anchor" href="#faceting"></a>
</h3>
<p>Each point-based plot can be split into multiple facets using the
options in the <code>"Facet"</code> checkbox. Users can facet by row
and/or column, using categorical factors in <code>colData(se)</code>
(for column-based plots) or <code>rowData(se)</code> (for row-based
plots). This provides a convenient way to stratify points in a single
plot by multiple factors of interest. Note that point selection can only
occur <em>within</em> a single facet at a time; points cannot be
selected across facets.</p>
</div>
<div class="section level3">
<h3 id="zooming-in-and-out">Zooming in and out<a class="anchor" aria-label="anchor" href="#zooming-in-and-out"></a>
</h3>
<p>Zooming in is possible by first selecting a region of interest in a
plot using the brush (drag and select); double-clicking on the brushed
area then zooms into the selected area. To zoom out to the original
plot, simply double-click at any location in the plot.</p>
</div>
</div>
<div class="section level2">
<h2 id="faq">FAQ<a class="anchor" aria-label="anchor" href="#faq"></a>
</h2>
<p><strong>Q: Can you implement a ‘Copy to clipboard’ button in the code
editor?</strong></p>
<p>A: This is not necessary, as one can click anywhere in the code
editor and instantly select all the code using a keyboard shortcut that
depends on your operating system.</p>
<p><strong>Q: When brushing with a transparency effect, it seems that
data points in the receiving plot are not made transparent/subsetted
correctly.</strong></p>
<p>A: What you see is an artefact of overplotting: in areas excessively
dense in points, transparency ceases to be an effective visual
effect.</p>
<p><strong>Q: Brushing on violin or square plots doesn’t seem to select
anything.</strong></p>
<p>A: For violin plots, points will be selected only if the brushed area
includes the center of the x-tick, i.e., the center of the violin plot.
This is intentional as it allows easy selection of all points in complex
grouped violin plots. Indeed, the location of a specific point on the
x-axis has no meaning. The same logic applies to the square plots, where
only the center of each square needs to be selected to obtain all the
points in the square.</p>
<p><strong>Q: I’d like to try <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> but I
can’t install it/I just want a quick peek. Is there something you can
do?</strong></p>
<p>A: We set up an instance of iSEE running on the <code>allen</code>
dataset at this address: <a href="http://shiny.imbei.uni-mainz.de:3838/iSEE" class="external-link uri">http://shiny.imbei.uni-mainz.de:3838/iSEE</a>. A range of
interactive tours showcasing a variety of data types is also available
here: <a href="https://github.com/iSEE/iSEE2018" class="external-link uri">https://github.com/iSEE/iSEE2018</a>. Please keep in mind
this is only for demonstration purposes, yet those instances show how
you or your system administrator can setup <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> for
analyzing and/or sharing your
<code>SummarizedExperiment</code>/<code>SingleCellExperiment</code>
precomputed object.</p>
<p><strong>Q: I would like to use <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> with my
<code>Seurat</code> object, how do I do?</strong></p>
<p>A: The <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> package
provides <code>as.SingleCellExperiment()</code> to coerce
<code>Seurat</code> objects to <code>SingleCellExperiment</code>
objects. This conversion includes assays, cell metadata, feature
metadata, and dimensionality reduction results. You can then use the
<code>SingleCellExperiment</code> object as usual.</p>
</div>
<div class="section level2">
<h2 id="additional-information">Additional information<a class="anchor" aria-label="anchor" href="#additional-information"></a>
</h2>
<p>Bug reports can be posted on the <a href="https://support.bioconductor.org" class="external-link">Bioconductor support site</a> or
raised as issues in the <em><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></em> GitHub repository. The
GitHub repository also contains the development version of the package,
where new functionality is added over time. The authors appreciate
well-considered suggestions for improvements or new features, or even
better, pull requests.</p>
<p>If you use <em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em> for
your analysis, please cite it as shown below:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html" class="external-link">citation</a></span><span class="op">(</span><span class="st">"iSEE"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; To cite package 'iSEE' in publications use:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   Rue-Albrecht K, Marini F, Soneson C, Lun ATL (2018). "iSEE:</span>
<span class="co">#&gt;   Interactive SummarizedExperiment Explorer." _F1000Research_, *7*,</span>
<span class="co">#&gt;   741. doi:10.12688/f1000research.14966.1</span>
<span class="co">#&gt;   &lt;https://doi.org/10.12688/f1000research.14966.1&gt;.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Article{,</span>
<span class="co">#&gt;     title = {iSEE: Interactive SummarizedExperiment Explorer},</span>
<span class="co">#&gt;     author = {Kevin Rue-Albrecht and Federico Marini and Charlotte Soneson and Aaron T. L. Lun},</span>
<span class="co">#&gt;     publisher = {F1000 Research, Ltd.},</span>
<span class="co">#&gt;     journal = {F1000Research},</span>
<span class="co">#&gt;     year = {2018},</span>
<span class="co">#&gt;     month = {Jun},</span>
<span class="co">#&gt;     volume = {7},</span>
<span class="co">#&gt;     pages = {741},</span>
<span class="co">#&gt;     doi = {10.12688/f1000research.14966.1},</span>
<span class="co">#&gt;   }</span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.2.0 (2022-04-22)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] scater_1.24.0               ggplot2_3.3.6              </span>
<span class="co">#&gt;  [3] scuttle_1.6.2               scRNAseq_2.10.0            </span>
<span class="co">#&gt;  [5] iSEE_2.9.0                  SingleCellExperiment_1.18.0</span>
<span class="co">#&gt;  [7] SummarizedExperiment_1.26.1 Biobase_2.56.0             </span>
<span class="co">#&gt;  [9] GenomicRanges_1.48.0        GenomeInfoDb_1.32.2        </span>
<span class="co">#&gt; [11] IRanges_2.30.0              S4Vectors_0.34.0           </span>
<span class="co">#&gt; [13] BiocGenerics_0.42.0         MatrixGenerics_1.8.0       </span>
<span class="co">#&gt; [15] matrixStats_0.62.0          BiocStyle_2.24.0           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] circlize_0.4.15               AnnotationHub_3.4.0          </span>
<span class="co">#&gt;   [3] BiocFileCache_2.4.0           systemfonts_1.0.4            </span>
<span class="co">#&gt;   [5] igraph_1.3.1                  lazyeval_0.2.2               </span>
<span class="co">#&gt;   [7] shinydashboard_0.7.2          splines_4.2.0                </span>
<span class="co">#&gt;   [9] BiocParallel_1.30.2           digest_0.6.29                </span>
<span class="co">#&gt;  [11] ensembldb_2.20.1              foreach_1.5.2                </span>
<span class="co">#&gt;  [13] htmltools_0.5.2               viridis_0.6.2                </span>
<span class="co">#&gt;  [15] fansi_1.0.3                   magrittr_2.0.3               </span>
<span class="co">#&gt;  [17] memoise_2.0.1                 ScaledMatrix_1.4.0           </span>
<span class="co">#&gt;  [19] cluster_2.1.3                 doParallel_1.0.17            </span>
<span class="co">#&gt;  [21] ComplexHeatmap_2.12.0         Biostrings_2.64.0            </span>
<span class="co">#&gt;  [23] pkgdown_2.0.3                 prettyunits_1.1.1            </span>
<span class="co">#&gt;  [25] colorspace_2.0-3              blob_1.2.3                   </span>
<span class="co">#&gt;  [27] rappdirs_0.3.3                ggrepel_0.9.1                </span>
<span class="co">#&gt;  [29] textshaping_0.3.6             xfun_0.31                    </span>
<span class="co">#&gt;  [31] dplyr_1.0.9                   crayon_1.5.1                 </span>
<span class="co">#&gt;  [33] RCurl_1.98-1.6                jsonlite_1.8.0               </span>
<span class="co">#&gt;  [35] iterators_1.0.14              glue_1.6.2                   </span>
<span class="co">#&gt;  [37] gtable_0.3.0                  zlibbioc_1.42.0              </span>
<span class="co">#&gt;  [39] XVector_0.36.0                GetoptLong_1.0.5             </span>
<span class="co">#&gt;  [41] DelayedArray_0.22.0           BiocSingular_1.12.0          </span>
<span class="co">#&gt;  [43] shape_1.4.6                   scales_1.2.0                 </span>
<span class="co">#&gt;  [45] DBI_1.1.2                     miniUI_0.1.1.1               </span>
<span class="co">#&gt;  [47] Rcpp_1.0.8.3                  progress_1.2.2               </span>
<span class="co">#&gt;  [49] viridisLite_0.4.0             xtable_1.8-4                 </span>
<span class="co">#&gt;  [51] clue_0.3-61                   rsvd_1.0.5                   </span>
<span class="co">#&gt;  [53] bit_4.0.4                     DT_0.23                      </span>
<span class="co">#&gt;  [55] htmlwidgets_1.5.4             httr_1.4.3                   </span>
<span class="co">#&gt;  [57] RColorBrewer_1.1-3            shinyAce_0.4.2               </span>
<span class="co">#&gt;  [59] ellipsis_0.3.2                XML_3.99-0.9                 </span>
<span class="co">#&gt;  [61] pkgconfig_2.0.3               sass_0.4.1                   </span>
<span class="co">#&gt;  [63] dbplyr_2.1.1                  utf8_1.2.2                   </span>
<span class="co">#&gt;  [65] tidyselect_1.1.2              rlang_1.0.2                  </span>
<span class="co">#&gt;  [67] later_1.3.0                   AnnotationDbi_1.58.0         </span>
<span class="co">#&gt;  [69] munsell_0.5.0                 BiocVersion_3.15.2           </span>
<span class="co">#&gt;  [71] tools_4.2.0                   cachem_1.0.6                 </span>
<span class="co">#&gt;  [73] cli_3.3.0                     ExperimentHub_2.4.0          </span>
<span class="co">#&gt;  [75] generics_0.1.2                RSQLite_2.2.14               </span>
<span class="co">#&gt;  [77] rintrojs_0.3.0                evaluate_0.15                </span>
<span class="co">#&gt;  [79] stringr_1.4.0                 fastmap_1.1.0                </span>
<span class="co">#&gt;  [81] yaml_2.3.5                    ragg_1.2.2                   </span>
<span class="co">#&gt;  [83] knitr_1.39                    bit64_4.0.5                  </span>
<span class="co">#&gt;  [85] fs_1.5.2                      purrr_0.3.4                  </span>
<span class="co">#&gt;  [87] AnnotationFilter_1.20.0       KEGGREST_1.36.0              </span>
<span class="co">#&gt;  [89] sparseMatrixStats_1.8.0       nlme_3.1-157                 </span>
<span class="co">#&gt;  [91] mime_0.12                     xml2_1.3.3                   </span>
<span class="co">#&gt;  [93] biomaRt_2.52.0                compiler_4.2.0               </span>
<span class="co">#&gt;  [95] beeswarm_0.4.0                interactiveDisplayBase_1.34.0</span>
<span class="co">#&gt;  [97] filelock_1.0.2                curl_4.3.2                   </span>
<span class="co">#&gt;  [99] png_0.1-7                     tibble_3.1.7                 </span>
<span class="co">#&gt; [101] bslib_0.3.1                   stringi_1.7.6                </span>
<span class="co">#&gt; [103] highr_0.9                     GenomicFeatures_1.48.1       </span>
<span class="co">#&gt; [105] desc_1.4.1                    lattice_0.20-45              </span>
<span class="co">#&gt; [107] ProtGenerics_1.28.0           Matrix_1.4-1                 </span>
<span class="co">#&gt; [109] shinyjs_2.1.0                 vctrs_0.4.1                  </span>
<span class="co">#&gt; [111] pillar_1.7.0                  lifecycle_1.0.1              </span>
<span class="co">#&gt; [113] BiocManager_1.30.18           jquerylib_0.1.4              </span>
<span class="co">#&gt; [115] GlobalOptions_0.1.2           BiocNeighbors_1.14.0         </span>
<span class="co">#&gt; [117] irlba_2.3.5                   bitops_1.0-7                 </span>
<span class="co">#&gt; [119] rtracklayer_1.56.0            httpuv_1.6.5                 </span>
<span class="co">#&gt; [121] BiocIO_1.6.0                  R6_2.5.1                     </span>
<span class="co">#&gt; [123] bookdown_0.26                 promises_1.2.0.1             </span>
<span class="co">#&gt; [125] gridExtra_2.3                 vipor_0.4.5                  </span>
<span class="co">#&gt; [127] codetools_0.2-18              colourpicker_1.1.1           </span>
<span class="co">#&gt; [129] assertthat_0.2.1              fontawesome_0.2.2            </span>
<span class="co">#&gt; [131] rprojroot_2.0.3               rjson_0.2.21                 </span>
<span class="co">#&gt; [133] withr_2.5.0                   shinyWidgets_0.7.0           </span>
<span class="co">#&gt; [135] GenomicAlignments_1.32.0      Rsamtools_2.12.0             </span>
<span class="co">#&gt; [137] GenomeInfoDbData_1.2.8        hms_1.1.1                    </span>
<span class="co">#&gt; [139] mgcv_1.8-40                   parallel_4.2.0               </span>
<span class="co">#&gt; [141] beachmat_2.12.0               grid_4.2.0                   </span>
<span class="co">#&gt; [143] DelayedMatrixStats_1.18.0     rmarkdown_2.14               </span>
<span class="co">#&gt; [145] Rtsne_0.16                    shiny_1.7.1                  </span>
<span class="co">#&gt; [147] ggbeeswarm_0.6.0              restfulr_0.0.13</span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-ganz2016rintrojs" class="csl-entry">
Ganz, C. 2016. <span>“<span class="nocase">rintrojs</span>: A Wrapper
for the Intro.js Library.”</span> <em>Journal of Open Source
Software</em> 1. <a href="http://dx.doi.org/10.21105/joss.00063" class="external-link">http://dx.doi.org/10.21105/joss.00063</a>.
</div>
<div id="ref-kra2018iSEE" class="csl-entry">
Rue-Albrecht, K., F. Marini, C. Soneson, and A. T. L. Lun. 2018.
<span>“iSEE: Interactive SummarizedExperiment Explorer.”</span>
<em>F1000Research</em> 7 (June): 741.
</div>
<div id="ref-tasic2016adult" class="csl-entry">
Tasic, B., V. Menon, T. N. Nguyen, T. K. Kim, T. Jarsky, Z. Yao, B.
Levi, et al. 2016. <span>“<span class="nocase"><span>A</span>dult mouse
cortical cell taxonomy revealed by single cell
transcriptomics</span>.”</span> <em>Nat. Neurosci.</em> 19 (2): 335–46.
</div>
<div id="ref-van2008visualizing" class="csl-entry">
Van der Maaten, L., and G. Hinton. 2008. <span>“Visualizing Data Using
t-<span>S</span><span>N</span><span>E</span>.”</span> <em>J. Mach.
Learn. Res.</em> 9 (2579-2605): 85.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Or an instance of a subclass, like a
<code>SingleCellExperiment</code> object.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Except when dealing with <a href="https://bioconductor.org/packages/3.15/iSEE/vignettes/custom.html" class="external-link">custom
panels</a>.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>That said, if there are categorical values for the
assayed values, these will be handled as described in the column data
plots.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>This is not effective if <code>se</code> does not
contain row names.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Federico Marini, Charlotte Soneson, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
