<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Describing the `ExperimentColorMap` class • iSEE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Describing the `ExperimentColorMap` class">
<meta property="og:description" content="iSEE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iSEE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic.html">An introduction to the `r Biocpkg("iSEE")` interface 
</a>
    </li>
    <li>
      <a href="../articles/bigdata.html">How to use `r Biocpkg("iSEE")` with big data
</a>
    </li>
    <li>
      <a href="../articles/configure.html">Configuring `r Biocpkg("iSEE")` apps
</a>
    </li>
    <li>
      <a href="../articles/custom.html">Deploying custom panels in the `iSEE` interface</a>
    </li>
    <li>
      <a href="../articles/ecm.html">Describing the `ExperimentColorMap` class</a>
    </li>
    <li>
      <a href="../articles/links.html">Sharing information across `r Biocpkg("iSEE")` panels
</a>
    </li>
    <li>
      <a href="../articles/voice.html">Controlling the `iSEE` interface using speech recognition</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iSEE/iSEE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="ecm_files/font-awesome-5.1.0/css/all.css" rel="stylesheet">
<link href="ecm_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Describing the <code>ExperimentColorMap</code>
class</h1>
                        <h4 data-toc-skip class="author">Kevin
Rue-Albrecht</h4>
            <address class="author_afil">
      MRC WIMM Centre for Computational Biology, University of Oxford,
Oxford, OX3 9DS,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Federico
Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), MainzCenter for Thrombosis and Hemostasis (CTH),
Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                              <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
            <address class="author_afil">
      Friedrich Miescher Institute for Biomedical Research, Basel,
SwitzerlandSIB Swiss Institute of
Bioinformatics<br><a class="author_email" href="mailto:#"></a><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Aaron
Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">29 June 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/iSEE/iSEE/blob/HEAD/vignettes/ecm.Rmd" class="external-link"><code>vignettes/ecm.Rmd</code></a></small>
      <div class="hidden name"><code>ecm.Rmd</code></div>

    </div>

    
    
<p><strong>Compiled date</strong>: 2022-06-29</p>
<p><strong>Last edited</strong>: 2018-03-08</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.15/iSEE" class="external-link">iSEE</a></em>
coordinates the coloration in every plot via the
<code>ExperimentColorMap</code> class <span class="citation">(Rue-Albrecht et al. 2018)</span>. Colors for samples
or features are defined from column or row metadata or assay values
using “colormaps”. Each colormap is a function that takes a single
integer argument and returns that number of distinct colors. The
<code>ExperimentColorMap</code> is a container that stores these
functions for use within the <code><a href="../reference/iSEE.html">iSEE()</a></code> function. Users can
define their own colormaps to customize coloration for specific assays
or covariates.</p>
</div>
<div class="section level2">
<h2 id="defining-colormaps">Defining colormaps<a class="anchor" aria-label="anchor" href="#defining-colormaps"></a>
</h2>
<div class="section level3">
<h3 id="colormaps-for-continuous-variables">Colormaps for continuous variables<a class="anchor" aria-label="anchor" href="#colormaps-for-continuous-variables"></a>
</h3>
<p>For continuous variables, the function will be asked to generate a
number of colors (21, by default). Interpolation will then be performed
internally to generate a color gradient. Users can use existing color
scales like <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis::viridis</a></code> or
<code>heat.colors</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Coloring for log-counts:</span></span>
<span><span class="va">logcounts_color_fun</span> <span class="op">&lt;-</span> <span class="fu">viridis</span><span class="fu">::</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span></span></code></pre></div>
<p>It is also possible to use a function that completely ignores any
arguments, and simply returns a fixed number of interpolation
points:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Coloring for FPKMs:</span></span>
<span><span class="va">fpkm_color_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"brown"</span>,<span class="st">"red"</span>,<span class="st">"orange"</span>,<span class="st">"yellow"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="colormaps-for-categorical-variables">Colormaps for categorical variables<a class="anchor" aria-label="anchor" href="#colormaps-for-categorical-variables"></a>
</h3>
<p>For categorical variables, the function should accept the number of
levels and return a color per level. Colors are automatically assigned
to factor levels in the specified order of the levels.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Coloring for the 'driver' metadata variable.</span></span>
<span><span class="va">driver_color_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="va">n</span>, <span class="st">"Set2"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Alternatively, the function can ignore its arguments and simply
return a named vector of colors if users want to specify the color for
each level explicitly It is the user’s responsibility to ensure that all
levels are accounted for<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. For instance, the following colormap
function will only be compatible with factors of two levels, namely
<code>"Y"</code> and <code>"N"</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Coloring for the QC metadata variable:</span></span>
<span><span class="va">qc_color_fun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">qc_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"forestgreen"</span>, <span class="st">"firebrick1"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">qc_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Y"</span>, <span class="st">"N"</span><span class="op">)</span></span>
<span>    <span class="va">qc_colors</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-colormap-hierarchy">The colormap hierarchy<a class="anchor" aria-label="anchor" href="#the-colormap-hierarchy"></a>
</h2>
<div class="section level3">
<h3 id="specific-and-shared-colormaps">Specific and shared colormaps<a class="anchor" aria-label="anchor" href="#specific-and-shared-colormaps"></a>
</h3>
<p>Colormaps can be defined by users at three different levels:</p>
<ul>
<li>Each individual assay, column data field, and row data field can be
assigned its own distinct colormap. Those colormaps are stored as named
lists of functions in the <code>assays</code>, <code>colData</code>, and
<code>rowData</code> slots, respectively, of the
<code>ExperimentColorMap</code>. This can be useful to easily remember
which assay is currently shown; to apply different color scale limits to
assays that vary on different ranges of values; or display boolean
information in an intuitive way, among many other scenarios.</li>
<li>
<em>Shared</em> colormaps can be defined for all assays, all column
data, and all row data. These colormaps are stored in the
<code>all_discrete</code> and <code>all_continuous</code> slots of the
<code>ExperimentColorMap</code>, as lists of functions named
<code>assays</code>, <code>colData</code>, and
<code>rowData</code>.</li>
<li>
<em>Global</em> colormaps can be defined for all categorical or
continuous data. Those two colormaps are stored in the
<code>global_discrete</code> and <code>global_continuous</code> slots of
the <code>ExperimentColorMap</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="searching-for-colors">Searching for colors<a class="anchor" aria-label="anchor" href="#searching-for-colors"></a>
</h3>
<p>When queried for a specific colormap of any type (assay, column data,
or row data), the following process takes place:</p>
<ul>
<li>A specific <em>individual</em> colormap is looked up in the
appropriate slot of the <code>ExperimentColorMap</code>.</li>
<li>If it is not found, the <em>shared</em> colormap of the appropriate
slot is looked up, according to whether the data are categorical or
continuous.</li>
<li>If it is not found, the <em>global</em> colormap is looked up,
according to whether the data are categorical or continuous.</li>
<li>If none of the above colormaps were defined, the
<code>ExperimentColorMap</code> will revert to the default
colormaps.</li>
</ul>
<p>By default, <code>viridis</code> is used as the default continuous
colormap, and <code>hcl</code> is used as the default categorical
colormap.</p>
</div>
</div>
<div class="section level2">
<h2 id="creating-the-experimentcolormap">Creating the <code>ExperimentColorMap</code><a class="anchor" aria-label="anchor" href="#creating-the-experimentcolormap"></a>
</h2>
<p>We store the set of colormap functions in an instance of the
<code>ExperimentColorMap</code> class. Named functions passed as
<code>assays</code>, <code>colData</code>, or <code>rowData</code>
arguments will be used for coloring data in those slots,
respectively.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></span><span class="op">)</span></span>
<span><span class="va">ecm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ExperimentColorMap-class.html">ExperimentColorMap</a></span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        counts <span class="op">=</span> <span class="va">heat.colors</span>,</span>
<span>        logcounts <span class="op">=</span> <span class="va">logcounts_color_fun</span>,</span>
<span>        cufflinks_fpkm <span class="op">=</span> <span class="va">fpkm_color_fun</span></span>
<span>    <span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        passes_qc_checks_s <span class="op">=</span> <span class="va">qc_color_fun</span>,</span>
<span>        driver_1_s <span class="op">=</span> <span class="va">driver_color_fun</span></span>
<span>    <span class="op">)</span>,</span>
<span>    all_continuous <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        assays <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">ecm</span></span>
<span><span class="co">#&gt; Class: ExperimentColorMap</span></span>
<span><span class="co">#&gt; assays(3): counts logcounts cufflinks_fpkm</span></span>
<span><span class="co">#&gt; colData(2): passes_qc_checks_s driver_1_s</span></span>
<span><span class="co">#&gt; rowData(0):</span></span>
<span><span class="co">#&gt; all_discrete(0):</span></span>
<span><span class="co">#&gt; all_continuous(1): assays</span></span></code></pre></div>
<p>Users can change the defaults for all assays or column data by
modifying the <em>shared</em> colormaps. Similarly, users can modify the
defaults for all continuous or categorical data by modifying the global
colormaps. This is demonstrated below for the continuous variables:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ExperimentColorMap-class.html">ExperimentColorMap</a></span><span class="op">(</span></span>
<span>    all_continuous<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="co"># shared</span></span>
<span>        assays<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">plasma</a></span>,</span>
<span>        colData<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">inferno</a></span></span>
<span>    <span class="op">)</span>,</span>
<span>    global_continuous<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="va"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span> <span class="co"># global</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Class: ExperimentColorMap</span></span>
<span><span class="co">#&gt; assays(0):</span></span>
<span><span class="co">#&gt; colData(0):</span></span>
<span><span class="co">#&gt; rowData(0):</span></span>
<span><span class="co">#&gt; all_discrete(0):</span></span>
<span><span class="co">#&gt; all_continuous(2): assays colData</span></span>
<span><span class="co">#&gt; global_continuous(1)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="benefits">Benefits<a class="anchor" aria-label="anchor" href="#benefits"></a>
</h2>
<p>The <code>ExperimentColorMap</code> class offers the following major
features:</p>
<ul>
<li>A single place to define flexible and lightweight sets of colormaps,
that may be saved and reused across sessions and projects outside the
app, to apply consistent coloring schemes across entire projects</li>
<li>A simple interface through accessors
<code>colDataColorMap(colormap, "coldata_name")</code> and setters
<code>assayColorMap(colormap, "assay_name") &lt;- colormap_function</code>
</li>
<li>An elegant fallback mechanism to consistently return a colormap,
even for undefined covariates, including a default categorical and
continuous colormap, respectively.</li>
<li>Three levels of colormaps override: individual, shared within slot
(i.e., <code>assays</code>, <code>colData</code>, <code>rowData</code>),
or shared globally between all categorical or continuous data
scales.</li>
</ul>
<p>Detailed examples on the use of <code>ExperimentColorMap</code>
objects are available in the documentation
<code><a href="../reference/ExperimentColorMap-class.html">?ExperimentColorMap</a></code>, as well as below.</p>
</div>
<div class="section level2">
<h2 id="demonstration">Demonstration<a class="anchor" aria-label="anchor" href="#demonstration"></a>
</h2>
<p>Here, we use the <code>allen</code> single-cell RNA-seq data set to
demonstrate the use of the <code>ExperimentColorMap</code> class. Using
the <code>sce</code> object that we created <a href="https://bioconductor.org/packages/3.15/iSEE/vignettes/basic.html" class="external-link">previously</a>,
we create an <code>iSEE</code> app with the
<code>SingleCellExperiment</code> object and the colormap generated
above.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iSEE.html">iSEE</a></span><span class="op">(</span><span class="va">sce</span>, colormap <span class="op">=</span> <span class="va">ecm</span><span class="op">)</span></span></code></pre></div>
<p>We run this using <code>runApp</code> to open the app on our
browser.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="va">app</span><span class="op">)</span></span></code></pre></div>
<p><img src="screenshots/ecm-demo.png" width="488"></p>
<p>Now, choose to color cells by <code>Column data</code> and select
<code>passes_qc_checks_s</code>. We will see that all cells that passed
QC (<code>Y</code>) are colored “forestgreen”, while the ones that
didn’t pass are colored firebrick.</p>
<p>If we color any plot by gene expression, we see that use of counts
follows the <code>heat.colors</code> coloring scheme; use of log-counts
follows the <code>viridis</code> coloring scheme; and use of FPKMs
follows the black-to-yellow scheme we defined in
<code>fpkm_color_fun</code>.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] iSEE_2.9.0                  SingleCellExperiment_1.18.0</span></span>
<span><span class="co">#&gt;  [3] SummarizedExperiment_1.26.1 Biobase_2.56.0             </span></span>
<span><span class="co">#&gt;  [5] GenomicRanges_1.48.0        GenomeInfoDb_1.32.2        </span></span>
<span><span class="co">#&gt;  [7] IRanges_2.30.0              S4Vectors_0.34.0           </span></span>
<span><span class="co">#&gt;  [9] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       </span></span>
<span><span class="co">#&gt; [11] matrixStats_0.62.0          BiocStyle_2.24.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] nlme_3.1-158           bitops_1.0-7           fs_1.5.2              </span></span>
<span><span class="co">#&gt;   [4] fontawesome_0.2.2      doParallel_1.0.17      RColorBrewer_1.1-3    </span></span>
<span><span class="co">#&gt;   [7] rprojroot_2.0.3        tools_4.2.0            bslib_0.3.1           </span></span>
<span><span class="co">#&gt;  [10] utf8_1.2.2             R6_2.5.1               DT_0.23               </span></span>
<span><span class="co">#&gt;  [13] vipor_0.4.5            mgcv_1.8-40            DBI_1.1.3             </span></span>
<span><span class="co">#&gt;  [16] colorspace_2.0-3       GetoptLong_1.0.5       tidyselect_1.1.2      </span></span>
<span><span class="co">#&gt;  [19] gridExtra_2.3          compiler_4.2.0         textshaping_0.3.6     </span></span>
<span><span class="co">#&gt;  [22] cli_3.3.0              shinyjs_2.1.0          desc_1.4.1            </span></span>
<span><span class="co">#&gt;  [25] DelayedArray_0.22.0    colourpicker_1.1.1     bookdown_0.27         </span></span>
<span><span class="co">#&gt;  [28] sass_0.4.1             scales_1.2.0           pkgdown_2.0.5         </span></span>
<span><span class="co">#&gt;  [31] systemfonts_1.0.4      stringr_1.4.0          digest_0.6.29         </span></span>
<span><span class="co">#&gt;  [34] rmarkdown_2.14         XVector_0.36.0         pkgconfig_2.0.3       </span></span>
<span><span class="co">#&gt;  [37] htmltools_0.5.2        highr_0.9              fastmap_1.1.0         </span></span>
<span><span class="co">#&gt;  [40] htmlwidgets_1.5.4      rlang_1.0.3            GlobalOptions_0.1.2   </span></span>
<span><span class="co">#&gt;  [43] shiny_1.7.1            shape_1.4.6            jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt;  [46] generics_0.1.2         jsonlite_1.8.0         dplyr_1.0.9           </span></span>
<span><span class="co">#&gt;  [49] RCurl_1.98-1.7         magrittr_2.0.3         GenomeInfoDbData_1.2.8</span></span>
<span><span class="co">#&gt;  [52] Matrix_1.4-1           Rcpp_1.0.8.3           munsell_0.5.0         </span></span>
<span><span class="co">#&gt;  [55] fansi_1.0.3            viridis_0.6.2          lifecycle_1.0.1       </span></span>
<span><span class="co">#&gt;  [58] rintrojs_0.3.0         stringi_1.7.6          yaml_2.3.5            </span></span>
<span><span class="co">#&gt;  [61] zlibbioc_1.42.0        grid_4.2.0             ggrepel_0.9.1         </span></span>
<span><span class="co">#&gt;  [64] promises_1.2.0.1       parallel_4.2.0         shinydashboard_0.7.2  </span></span>
<span><span class="co">#&gt;  [67] crayon_1.5.1           miniUI_0.1.1.1         lattice_0.20-45       </span></span>
<span><span class="co">#&gt;  [70] splines_4.2.0          circlize_0.4.15        knitr_1.39            </span></span>
<span><span class="co">#&gt;  [73] ComplexHeatmap_2.12.0  pillar_1.7.0           igraph_1.3.2          </span></span>
<span><span class="co">#&gt;  [76] rjson_0.2.21           codetools_0.2-18       glue_1.6.2            </span></span>
<span><span class="co">#&gt;  [79] evaluate_0.15          BiocManager_1.30.18    httpuv_1.6.5          </span></span>
<span><span class="co">#&gt;  [82] png_0.1-7              vctrs_0.4.1            foreach_1.5.2         </span></span>
<span><span class="co">#&gt;  [85] gtable_0.3.0           purrr_0.3.4            clue_0.3-61           </span></span>
<span><span class="co">#&gt;  [88] assertthat_0.2.1       cachem_1.0.6           ggplot2_3.3.6         </span></span>
<span><span class="co">#&gt;  [91] xfun_0.31              mime_0.12              xtable_1.8-4          </span></span>
<span><span class="co">#&gt;  [94] later_1.3.0            ragg_1.2.2             viridisLite_0.4.0     </span></span>
<span><span class="co">#&gt;  [97] tibble_3.1.7           iterators_1.0.14       memoise_2.0.1         </span></span>
<span><span class="co">#&gt; [100] cluster_2.1.3          shinyWidgets_0.7.0     ellipsis_0.3.2        </span></span>
<span><span class="co">#&gt; [103] shinyAce_0.4.2</span></span>
<span><span class="co"># devtools::session_info()</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-kra2018iSEE" class="csl-entry">
Rue-Albrecht, K., F. Marini, C. Soneson, and A. T. L. Lun. 2018.
<span>“iSEE: Interactive SummarizedExperiment Explorer.”</span>
<em>F1000Research</em> 7 (June): 741.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Needless to say, these functions should not be used as
shared or global colormaps.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Federico Marini, Charlotte Soneson, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
