<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sharing information across `r Biocpkg("iSEE")` panels
 • iSEE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sharing information across `r Biocpkg(" isee panels>
<meta property="og:description" content="iSEE">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iSEE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basic.html">An introduction to the `r Biocpkg("iSEE")` interface 
</a>
    </li>
    <li>
      <a href="../articles/bigdata.html">How to use `r Biocpkg("iSEE")` with big data
</a>
    </li>
    <li>
      <a href="../articles/configure.html">Configuring `r Biocpkg("iSEE")` apps
</a>
    </li>
    <li>
      <a href="../articles/custom.html">Deploying custom panels in the `iSEE` interface</a>
    </li>
    <li>
      <a href="../articles/ecm.html">Describing the `ExperimentColorMap` class</a>
    </li>
    <li>
      <a href="../articles/links.html">Sharing information across `r Biocpkg("iSEE")` panels
</a>
    </li>
    <li>
      <a href="../articles/voice.html">Controlling the `iSEE` interface using speech recognition</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iSEE/iSEE/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="links_files/font-awesome-5.1.0/css/all.css" rel="stylesheet">
<link href="links_files/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip></h1>
<p>Sharing information across <em><a href="https://bioconductor.org/packages/3.16/iSEE" class="external-link">iSEE</a></em>
panels</p>
                        <h4 data-toc-skip class="author">Kevin
Rue-Albrecht</h4>
            <address class="author_afil">
      MRC WIMM Centre for Computational Biology, University of Oxford,
Oxford, OX3 9DS,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Federico
Marini</h4>
            <address class="author_afil">
      Institute of Medical Biostatistics, Epidemiology and Informatics
(IMBEI), MainzCenter for Thrombosis and Hemostasis (CTH),
Mainz<br><a class="author_email" href="mailto:#"></a><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a>
      </address>
                              <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
            <address class="author_afil">
      Friedrich Miescher Institute for Biomedical Research, Basel,
SwitzerlandSIB Swiss Institute of
Bioinformatics<br><a class="author_email" href="mailto:#"></a><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Aaron
Lun</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:infinite.monkeys.with.keyboards@gmail.com" class="email">infinite.monkeys.with.keyboards@gmail.com</a>
      
                  
            <h4 data-toc-skip class="date">24 July 2022</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/iSEE/iSEE/blob/HEAD/vignettes/links.Rmd" class="external-link"><code>vignettes/links.Rmd</code></a></small>
      <div class="hidden name"><code>links.Rmd</code></div>

    </div>

    
    
<p><strong>Compiled date</strong>: 2022-07-24</p>
<p><strong>Last edited</strong>: 2020-04-20</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<p><img src="../../../_temp/Library/iSEE/www/iSEE.png" width="50%" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>One of <em><a href="https://bioconductor.org/packages/3.16/iSEE" class="external-link">iSEE</a></em>’s main
features is the ability to share information between panels. This
facilitates deeper exploration of a dataset by allowing users to
visualize relationships across multiple metrics. <code><a href="../reference/iSEE.html">iSEE()</a></code>
currently supports three modes of information sharing between panels -
multiple selections, single selections and dynamic selections - which
are demonstrated in this vignette using the Allen dataset processed in
the <a href="https://bioconductor.org/packages/3.16/iSEE/vignettes/basic.html" class="external-link">previous
vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="multiple-selections">Multiple selections<a class="anchor" aria-label="anchor" href="#multiple-selections"></a>
</h2>
<div class="section level3">
<h3 id="basic-use">Basic use<a class="anchor" aria-label="anchor" href="#basic-use"></a>
</h3>
<p>As its name suggests, this involves selecting multiple features or
samples in one panel and transmitting their identities to another panel
to affect the visualization. To demonstrate, we will create a small app
involving a single reduced dimension plot and a column metadata
plot.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></span><span class="op">)</span>
<span class="va">app</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iSEE.html">iSEE</a></span><span class="op">(</span><span class="va">sce</span>, initial<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    <span class="fu"><a href="../reference/ReducedDimensionPlot-class.html">ReducedDimensionPlot</a></span><span class="op">(</span><span class="op">)</span>,
    <span class="fu"><a href="../reference/ColumnDataPlot-class.html">ColumnDataPlot</a></span><span class="op">(</span><span class="op">)</span>
<span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="screenshots/links-multi-naive.png" width="488"></p>
<p>We indicate that we want one panel to “receive” a multiple selection
from another panel. This is done by specifying the selection source (for
samples/columns in this case, given the nature of the two panels) and
indicating that the column data plot is to receive a selection from the
reduced dimension plot.</p>
<p><img src="screenshots/links-multi-unselected.png" width="488"></p>
<p>We can click-and-drag to select multiple points in the reduced
dimension plot with a brush, which highlights those same points in the
column data plot. This enables users to easily explore relationships
between different visualizations in the <code><a href="../reference/iSEE.html">iSEE()</a></code>
interface.</p>
<p><img src="screenshots/links-multi-brushed.png" width="488"></p>
<p>Alternatively, a single click will lay down lasso waypoints for a
non-rectangular selection. Once closed, the points in the lasso will be
transmitted to the column data plot.</p>
<p><img src="screenshots/links-multi-lasso.png" width="488"></p>
</div>
<div class="section level3">
<h3 id="selection-effects">Selection effects<a class="anchor" aria-label="anchor" href="#selection-effects"></a>
</h3>
<p>Transparency is the default aesthetic effect used to distinguish
points in multiple selections in receiving plots. Alternatively, we may
use color:</p>
<p><img src="screenshots/links-multi-color.png" width="488"></p>
<p>Another option is to restrict the receiving plot so that it only
shows the points in the multiple selection. This effectively “gates” the
dataset on the selection in the reduced dimension plot, analogous to
identification populations of interest in flow cytometry studies.
Indeed, this gating process can be repeated as many times as desired; a
multiple selection could made on the column data plot and transmitted to
another panel, and so on. (See <a href="https://github.com/iSEE/iSEE2018/blob/master/tours/cytof/app.R" class="external-link">this
mass cytometry tour</a> for an example.)</p>
<p><img src="screenshots/links-multi-restrict.png" width="488"></p>
<p>Of course, not all receiving panels need to be plots. If we transmit
a multiple selection to a column data table, the effect is to subset the
rows of the table corresponding to the selected points.</p>
<p><img src="screenshots/links-multi-table.png" width="488"></p>
<p>As an aside, it is equally possible for a table to transmit a
multiple selection to other panels. This is achieved using the search
fields to subset the dataset to the desired selection.</p>
<p><img src="screenshots/links-multi-table2.png" width="488"></p>
</div>
<div class="section level3">
<h3 id="saving-selections">Saving selections<a class="anchor" aria-label="anchor" href="#saving-selections"></a>
</h3>
<p>In certain panels (usually plots), multiple selections can be saved
to form disjoint selections. Receiving plots will respond to the union
of all active and saved selections:</p>
<p><img src="screenshots/links-multi-saved.png" width="488"></p>
<p>The saved selection history operates on a first-in-last-out basis.
Upon saving, a snapshot is taken of the current “active” selection,
i.e., the brush or lasso that is just created. Deletion will only
operate on the last saved selection.</p>
</div>
</div>
<div class="section level2">
<h2 id="single-selections">Single selections<a class="anchor" aria-label="anchor" href="#single-selections"></a>
</h2>
<p>Another mode of information sharing involves transmitting a selection
of a single feature or sample. This allows users to conveniently direct
other panels to focus on a feature or sample of interest. For example,
we can transmit a single selection from a row data table to a reduced
dimension plot, instructing the latter to color points by the expression
of the chosen feature.</p>
<p><img src="screenshots/links-single-color.png" width="488"></p>
<p>The same approach can be used to control what is plotted on a feature
assay plot. Clicking on a different row of the table will directly
change the axes (in this case, the y-axis) of the plot, allowing the
user to synchronise different aspects of the <code><a href="../reference/iSEE.html">iSEE()</a></code>
interface to whatever is currently of interest.</p>
<p><img src="screenshots/links-single-yaxis.png" width="488"></p>
<p>We can also perform single selections on sample identities. In the
example below, the reduced dimension plot highlights the location of the
sample chosen in the column data table. This is useful for checking the
behavior of specific samples of interest, e.g., during quality
control.</p>
<p><img src="screenshots/links-single-sample.png" width="488"></p>
<p>Furthermore, it is possible to transmit single selections from a plot
using brushes or lassos. If the brush/lasso contains multiple points,
one of them is arbitrarily chosen for the purposes of obtaining a single
selection. Below, we select a single highly variable gene to examine its
distribution of expression values across the reduced dimension plot.</p>
<p><img src="screenshots/links-single-plot2plot.png" width="488"></p>
</div>
<div class="section level2">
<h2 id="dynamic-selections">Dynamic selections<a class="anchor" aria-label="anchor" href="#dynamic-selections"></a>
</h2>
<p>Most panels will have an option to dynamically change the choice of
transmitting panel according to the last active selection in the app.
This allows users to, for example, simply brush on any plot in the app
and have all participating receiving panels immediately use that
selection without requiring manual resetting of the transmitter. To
illustrate, let’s set up an instance with three different plots that all
represent points as samples.</p>
<p><img src="screenshots/links-dynamic-multi.png" width="488"></p>
<p>Notice how we have checked the dynamic source selection option in the
selection parameter box for all panels. This means that, upon making a
selection in any one of the plots, all of the other plots will
automatically respond as if they had been manually set to receive a
transmission from that plot.</p>
<p><img src="screenshots/links-dynamic-multi2.png" width="488"></p>
<p>The same logic applies for some parameters that respond to single
selections. This allows a panel to respond to an appropriate single
selection from any other panel in the interface, without requiring the
user to manually set the relationship between panels. In the example
below, users can easily define the y-axis of the feature assay plot from
a selected row in the row data table or from a selected single point in
the row data plot.</p>
<p><img src="screenshots/links-dynamic-single.png" width="488"></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 4.2.0 (2022-04-22)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">#&gt; [8] base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] iSEE_2.9.4                  SingleCellExperiment_1.19.0</span>
<span class="co">#&gt;  [3] SummarizedExperiment_1.27.1 Biobase_2.57.1             </span>
<span class="co">#&gt;  [5] GenomicRanges_1.49.0        GenomeInfoDb_1.33.3        </span>
<span class="co">#&gt;  [7] IRanges_2.31.0              S4Vectors_0.35.1           </span>
<span class="co">#&gt;  [9] BiocGenerics_0.43.0         MatrixGenerics_1.9.1       </span>
<span class="co">#&gt; [11] matrixStats_0.62.0          BiocStyle_2.25.0           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] colorspace_2.0-3       rjson_0.2.21           ellipsis_0.3.2        </span>
<span class="co">#&gt;   [4] rprojroot_2.0.3        circlize_0.4.15        XVector_0.37.0        </span>
<span class="co">#&gt;   [7] GlobalOptions_0.1.2    fs_1.5.2               clue_0.3-61           </span>
<span class="co">#&gt;  [10] ggrepel_0.9.1          DT_0.23                fansi_1.0.3           </span>
<span class="co">#&gt;  [13] codetools_0.2-18       splines_4.2.0          doParallel_1.0.17     </span>
<span class="co">#&gt;  [16] cachem_1.0.6           knitr_1.39             jsonlite_1.8.0        </span>
<span class="co">#&gt;  [19] cluster_2.1.3          png_0.1-7              shinydashboard_0.7.2  </span>
<span class="co">#&gt;  [22] shiny_1.7.1            BiocManager_1.30.18    compiler_4.2.0        </span>
<span class="co">#&gt;  [25] assertthat_0.2.1       Matrix_1.4-1           fastmap_1.1.0         </span>
<span class="co">#&gt;  [28] cli_3.3.0              later_1.3.0            htmltools_0.5.2       </span>
<span class="co">#&gt;  [31] tools_4.2.0            igraph_1.3.2           gtable_0.3.0          </span>
<span class="co">#&gt;  [34] glue_1.6.2             GenomeInfoDbData_1.2.8 dplyr_1.0.9           </span>
<span class="co">#&gt;  [37] Rcpp_1.0.8.3           jquerylib_0.1.4        pkgdown_2.0.6.9000    </span>
<span class="co">#&gt;  [40] vctrs_0.4.1            nlme_3.1-158           iterators_1.0.14      </span>
<span class="co">#&gt;  [43] rintrojs_0.3.0         xfun_0.31              stringr_1.4.0         </span>
<span class="co">#&gt;  [46] mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.1       </span>
<span class="co">#&gt;  [49] shinyAce_0.4.2         zlibbioc_1.43.0        scales_1.2.0          </span>
<span class="co">#&gt;  [52] colourpicker_1.1.1     ragg_1.2.2             promises_1.2.0.1      </span>
<span class="co">#&gt;  [55] parallel_4.2.0         RColorBrewer_1.1-3     ComplexHeatmap_2.13.0 </span>
<span class="co">#&gt;  [58] yaml_2.3.5             memoise_2.0.1          ggplot2_3.3.6         </span>
<span class="co">#&gt;  [61] sass_0.4.1             stringi_1.7.6          highr_0.9             </span>
<span class="co">#&gt;  [64] desc_1.4.1             foreach_1.5.2          shape_1.4.6           </span>
<span class="co">#&gt;  [67] rlang_1.0.3            pkgconfig_2.0.3        systemfonts_1.0.4     </span>
<span class="co">#&gt;  [70] bitops_1.0-7           evaluate_0.15          lattice_0.20-45       </span>
<span class="co">#&gt;  [73] fontawesome_0.2.2      purrr_0.3.4            htmlwidgets_1.5.4     </span>
<span class="co">#&gt;  [76] tidyselect_1.1.2       magrittr_2.0.3         bookdown_0.27         </span>
<span class="co">#&gt;  [79] R6_2.5.1               generics_0.1.2         DelayedArray_0.23.0   </span>
<span class="co">#&gt;  [82] DBI_1.1.3              pillar_1.7.0           mgcv_1.8-40           </span>
<span class="co">#&gt;  [85] RCurl_1.98-1.7         tibble_3.1.7           crayon_1.5.1          </span>
<span class="co">#&gt;  [88] shinyWidgets_0.7.0     utf8_1.2.2             rmarkdown_2.14        </span>
<span class="co">#&gt;  [91] GetoptLong_1.0.5       grid_4.2.0             digest_0.6.29         </span>
<span class="co">#&gt;  [94] xtable_1.8-4           httpuv_1.6.5           textshaping_0.3.6     </span>
<span class="co">#&gt;  [97] munsell_0.5.0          viridisLite_0.4.0      vipor_0.4.5           </span>
<span class="co">#&gt; [100] bslib_0.3.1            shinyjs_2.1.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Federico Marini, Charlotte Soneson, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
